# 2.2 文本分词

In [1]:
# 导入所需库：os用于文件路径判断和操作，urllib.request用于网络请求下载文件
import os
import urllib.request

# 定义数据集保存路径和下载链接
file_name = "the-verdict.txt"  # 本地保存的文件名
url = "https://raw.githubusercontent.com/rasbt/LLMs-from-scratch/main/ch02/01_main-chapter-code/the-verdict.txt"  # GitHub上的原始文本链接

# 先判断本地是否已经存在该文件，避免重复下载
if not os.path.exists(file_name):
    # 发起网络请求下载文件
    urllib.request.urlretrieve(url, file_name)
    print(f"数据集已成功下载至：{os.path.abspath(file_name)}")
else:
    print(f"本地已存在{file_name}，无需重复下载")

本地已存在the-verdict.txt，无需重复下载


In [2]:
# 以只读模式（"r"）打开文本文件，指定编码格式为utf-8（避免中文/特殊字符乱码，Windows系统尤为重要）
with open("the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()  # 读取文件全部内容，保存到raw_text变量中

# 验证读取结果
# 1. 打印前200个字符，查看文本是否正常读取，无乱码
print("原始文本前200个字符：")
print(raw_text[:200])
# 2. 统计文本总字符数，验证是否完整读取（正常应为约20479个字符）
total_chars = len(raw_text)
print(f"\n原始文本总字符数：{total_chars}")

原始文本前200个字符：
I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no great surprise to me to hear that, in the height of his glory, he had dropped his painting, married a

原始文本总字符数：20479


In [3]:
import re  # 导入正则表达式库

# 定义测试文本
test_text = "Hello, world. This, is a test."

# 使用re.split()进行拆分，规则r'(\s)'：匹配所有空白字符（空格、制表符等），并保留分隔符
token_list = re.split(r'(\s)', test_text)

# 打印拆分结果
print("基础版（按空格拆分）分词结果：")
print(token_list)
# 输出结果：['Hello,', ' ', 'world.', ' ', 'This,', ' ', 'is', ' ', 'a', ' ', 'test.']

基础版（按空格拆分）分词结果：
['Hello,', ' ', 'world.', ' ', 'This,', ' ', 'is', ' ', 'a', ' ', 'test.']


In [4]:
import re

test_text = "Hello, world. This, is a test."

# 正则规则r'([,.]|\s)'：匹配逗号（,）、句号（.）或空白字符（\s），并保留这些分隔符
token_list = re.split(r'([,.]|\s)', test_text)

# 过滤空字符串和纯空白字符（拆分后可能出现空元素，需要清理）
clean_tokens = [token for token in token_list if token.strip()]

# 打印清理后的分词结果
print("进阶版（拆分空格+逗号+句号）分词结果：")
print(clean_tokens)
# 输出结果：['Hello', ',', 'world', '.', 'This', ',', 'is', 'a', 'test', '.']

进阶版（拆分空格+逗号+句号）分词结果：
['Hello', ',', 'world', '.', 'This', ',', 'is', 'a', 'test', '.']


In [5]:
import re

# 定义包含更多特殊字符的测试文本
test_text = "Hello, world. Is this-- a test? \"Yes, it is.\""

# 正则规则：匹配逗号、分号、问号、感叹号、句号、引号、括号、双横线或空白字符
pattern = r'([,;?!."()\']|--|\s)'
token_list = re.split(pattern, test_text)

# 过滤空字符串和纯空白字符
clean_tokens = [token.strip() for token in token_list if token.strip()]

# 打印分词结果
print("通用版（支持更多标点和特殊字符）分词结果：")
print(clean_tokens)
# 输出结果：['Hello', ',', 'world', '.', 'Is', 'this', '--', 'a', 'test', '?', '"', 'Yes', ',', 'it', 'is', '.', '"']

通用版（支持更多标点和特殊字符）分词结果：
['Hello', ',', 'world', '.', 'Is', 'this', '--', 'a', 'test', '?', '"', 'Yes', ',', 'it', 'is', '.', '"']


In [6]:
import re
import os

# 1. 先读取本地数据集（若未读取，先执行读取代码）
if not os.path.exists("the-verdict.txt"):
    print("请先下载数据集！")
else:
    with open("the-verdict.txt", "r", encoding="utf-8") as f:
        raw_text = f.read()

    # 2. 定义通用分词正则规则（与参考代码保持一致，包含,:;?_!"()'、--和空白符）
    # 补充了参考代码中的 : 和 _ 符号，保持分词规则统一
    pattern = r'([,.:;?_!"()\']|--|\s)'

    # 3. 执行分词
    token_list = re.split(pattern, raw_text)

    # 4. 过滤空字符串和纯空白字符，得到预处理后的Token列表（与参考代码逻辑一致）
    preprocessed_tokens = [token.strip() for token in token_list if token.strip()]

    # 5. 验证分词结果（调整输出前N个Token的数量为30，与参考代码一致）
    print(f"原始文本总字符数：{len(raw_text)}")
    print(f"分词后总Token数：{len(preprocessed_tokens)}")
    print(f"前30个Token：{preprocessed_tokens[:30]}")

原始文本总字符数：20479
分词后总Token数：4690
前30个Token：['I', 'HAD', 'always', 'thought', 'Jack', 'Gisburn', 'rather', 'a', 'cheap', 'genius', '--', 'though', 'a', 'good', 'fellow', 'enough', '--', 'so', 'it', 'was', 'no', 'great', 'surprise', 'to', 'me', 'to', 'hear', 'that', ',', 'in']


# 总结

1. 2.1 节词嵌入：核心是 “文字转数值向量”，是大模型理解文本的基础，语义相近的文本向量距离更近，为后续模型训练提供输入支撑；

2. 2.2 节文本分词：核心是 “文本拆分为 Token”，是词嵌入的前置步骤，我们通过正则表达式实现了从简单到复杂的分词，并用《The Verdict》数据集完成了实操验证；

3. 核心流程梳理：原始文本 → 分词（拆分为 Token） → 转换为 Token ID → 词嵌入（转换为数值向量） → 输入 LLM 训练，这是后续我们构建大模型的核心数据处理流程。


# 2.3 将令牌转换为令牌ID

In [7]:
all_words = sorted(set(preprocessed_tokens))
vocab_size  = len(all_words)
vocab_size

1130

In [8]:
vocab = {token: integer for integer, token in enumerate(all_words)}

In [9]:
vocab["Jack"]

57

In [10]:
int_to_str = {i: s for s, i in vocab.items()}

In [11]:
int_to_str[57]

'Jack'

In [12]:
import re

class SimpleTokenizerV1:
    def __init__(self, vocab):
        self.str_to_int = vocab
        self.int_to_str = {i: s for s, i in vocab.items()}
    
    def encode(self, text):
        preprocessed = re.split(r'([,;:.?!()"\\\'-]|\s|--)', text)
        preprocessed = [item.strip() for item in preprocessed if item.strip()]
        ids = [self.str_to_int[s] for s in preprocessed]
        return ids
    
    def decode(self, ids):
        
        text = " ".join([self.int_to_str[i] for i in ids])
        text = re.sub(r'\s+([,;:.?!()"\\\'])', r'\1', text)
        return text

In [13]:
tokenizer = SimpleTokenizerV1(vocab)

text = """"It's the last he painted, you know," Mrs. Gisburn said with pardonable pride."""

ids = tokenizer.encode(text)

print(ids)

[1, 56, 2, 850, 988, 602, 533, 746, 5, 1126, 596, 5, 1, 67, 7, 38, 851, 1108, 754, 793, 7]


In [14]:
tokenizer.decode(ids)

'" It\' s the last he painted, you know," Mrs. Gisburn said with pardonable pride.'

In [15]:
tokenizer.decode(tokenizer.encode(text))

'" It\' s the last he painted, you know," Mrs. Gisburn said with pardonable pride.'

# 总结

1、去重：使用 set(preprocessed) 剔除重复 Token

2、排序：使用 sorted() 函数对去重后的 Token 按字母顺序排列

3、用 enumerate() 函数遍历排序后的 Token 列表，生成映射关系，为每个 Token 分配递增的唯一整数 ID，形成 Token → ID 的正向映射

4、对字典进行反向，形成 ID → Token的反向映射

5、简单分词器（SimpleTokenizer）：正向映射：字符串 Token → 整数 ID，反向映射：整数 ID → 字符串 Token

6、encode 方法：基于正向映射，将 Token 列表转换为 Token ID 序列

7、decode 方法：基于反向映射，将 Token ID 序列还原为文本字符串

# 2.4 添加特殊上下文令牌

In [16]:
tokenizer = SimpleTokenizerV1(vocab)

# text = "Hello, do you like tea. Is this-- a test?"

# tokenizer.encode(text)

-上面产生一个错误，因为单词“Hello”不包含在词汇表中

-为了处理这种情况，我们可以在词汇表中添加特殊的标记，如`“<|unk|>”`来表示未知的单词

-由于我们已经在扩展词汇表，让我们添加另一个名为`”<|endoftext|>”`的标记，该标记在GPT-2训练中用于表示文本的结尾（它也用于连接文本之间，例如如果我们的训练数据集由多篇文章、书籍等组成）

In [17]:
all_tokens = sorted(list(set(preprocessed_tokens)))
all_tokens.extend(["<|endoftext|>", "<|unk|>"])

vocab = {token:integer for integer, token in enumerate(all_tokens)}

In [18]:
len(vocab.items())

1132

In [19]:
for i, item in enumerate(list(vocab.items())[-5:]):
    print(item)

('younger', 1127)
('your', 1128)
('yourself', 1129)
('<|endoftext|>', 1130)
('<|unk|>', 1131)


In [20]:
class SimpleTokenizerV2:
    def __init__(self, vocab):
        self.str_to_int = vocab
        self.int_to_str = {i: s for s, i in vocab.items()}
    
    def encode(self, text):
        preprocessed = re.split(r'([,;:.?!()"\\\'-]|\s|--)', text)
        preprocessed = [item.strip() for item in preprocessed if item.strip()]
        preprocessed = [
            item if item in self.str_to_int 
            else "<|unk|>" for item in preprocessed
            ]
        ids = [self.str_to_int[s] for s in preprocessed]
        return ids
    
    def decode(self, ids):
        
        text = " ".join([self.int_to_str[i] for i in ids])
        text = re.sub(r'\s+([,;:.?!()"\\\'])', r'\1', text)
        return text

In [21]:
tokenizer = SimpleTokenizerV2(vocab)

text1 = "Hello, do you like tea?"
text2 = "In the sunlit terraces of the palace."

text = " <|endoftext|> ".join((text1, text2))

print(text)

Hello, do you like tea? <|endoftext|> In the sunlit terraces of the palace.


In [22]:
tokenizer.encode(text)

[1131, 5, 355, 1126, 628, 975, 10, 1130, 55, 988, 956, 984, 722, 988, 1131, 7]

In [23]:
tokenizer.decode((tokenizer.encode(text)))

'<|unk|>, do you like tea? <|endoftext|> In the sunlit terraces of the <|unk|>.'

# 总结

- 拼接多个独立文本时，用 `<endoftext>` 分隔（比如把不同文章连起来训练时，用它区分文本边界）；

- 扩展词汇表时，会按需加特殊 token（比如先加 `<unk>` 处理未知词，再加 `<endoftext>` 处理文本结尾）。

# 2.5 字节对编码

- GPT-2 使用字节对编码（BPE）作为其分词器。
- 它允许模型将不在预定义词汇表中的单词拆分为更小的子词单元，甚至是单个字符，从而能够处理未登录词（out-of-vocabulary words）。
- 例如，如果 GPT-2 的词汇表中没有 “unfamiliarword” 这个词，它可能会将其分词为 `["unfam", "iliar", "word"]` 或其他子词拆分形式，具体取决于它训练得到的 BPE 合并规则。
- 原始的 BPE 分词器可以在这里找到：https://github.com/openai/gpt-2/blob/master/src/encoder.py
- 在本章中，我们使用 OpenAI 开源的 `tiktoken` 库中的 BPE 分词器，该库的核心算法用 Rust 实现，以提高计算性能。

In [24]:
# 导入 tiktoken 库，这是 OpenAI 开源的高性能分词器
import tiktoken

# 导入 importlib 模块，用于获取已安装库的版本信息
import importlib

# 打印当前安装的 tiktoken 版本号
print("tiktoken version:", importlib.metadata.version("tiktoken"))

tiktoken version: 0.7.0


In [25]:
# 获取 GPT-2 对应的 BPE 分词器实例
# 'gpt2' 表示使用与 GPT-2 完全相同的分词规则和词汇表
tokenizer = tiktoken.get_encoding('gpt2')

In [26]:
# 定义要处理的文本
# <|endoftext|> 是 GPT-2 使用的特殊结束标记，用来分隔不同的文本片段
text = (
    "Hello, do you like tea? <|endoftext|> In the sunlit terraces"
    "of someunknownPlace."
)

integers = tokenizer.encode(text, allowed_special={'<|endoftext|>'})

integers

[15496,
 11,
 466,
 345,
 588,
 8887,
 30,
 220,
 50256,
 554,
 262,
 4252,
 18250,
 8812,
 2114,
 1659,
 617,
 34680,
 27271,
 13]

In [27]:
# 将整数 Token ID 解码回原始文本
strings = tokenizer.decode(integers)

strings

'Hello, do you like tea? <|endoftext|> In the sunlit terracesof someunknownPlace.'

![](images/bpe分词器原理.png)

BPE 的原理可以概括为：

1. **基础底裤**：自带 256 个单个字符的基础 Token（覆盖所有字节）；
2. **学高频对**：训练时合并文本里的高频字符对，生成子词 Token；
3. **拆未知词**：遇到没见过的词，就用已学的子词 + 基础字符‘拼着拆’，保证能处理任意文本。

1. **未登录词处理**：BPE 通过 “拆为基础字节 + 反向合并已知子词” 识别未知词，核心是 “最小单元兜底，高频组合优化”。
2. **BPE 核心逻辑**：训练时合并高频字节对构建词汇表，编码时按规则合并，解码时反向拆分。
3. **实用选择**：教学用自定义 BPE 实现理解原理，实际项目用`tiktoken`保证性能。

# 2.6 使用滑动窗口进行数据采样

In [28]:
# 以只读模式打开文本文件，编码为UTF-8
with open("the-verdict.txt", "r", encoding="utf-8") as f:
    # 读取文件中的全部原始文本
    raw_text = f.read()

In [29]:
# 使用分词器将原始文本编码为模型可处理的数字ID序列
enc_text = tokenizer.encode(raw_text)
# 打印编码后的序列长度，即文本被切分成的token数量
print(len(enc_text))

5145


In [30]:
# 从编码后的序列中截取从第50个token开始的片段作为示例
enc_sample = enc_text[50:]

In [31]:
# 设置上下文窗口大小为4
context_size = 4

# 取enc_sample的前context_size个token作为输入x
x = enc_sample[:context_size]
# 取enc_sample的第1到context_size+1个token作为目标y（输入向右偏移1位）
y = enc_sample[1:context_size+1]

# 打印输入x和目标y的token序列
print(f"x: {x}")
print(f"y: {y}")

x: [290, 4920, 2241, 287]
y: [4920, 2241, 287, 257]


In [32]:
# 遍历1到context_size（包含）的范围，展示逐步预测过程
for i in range(1, context_size+1):
    # 取前i个token作为当前上下文
    context = enc_sample[:i]
    # 第i个token是当前上下文需要预测的目标词
    desired = enc_sample[i]

    # 打印上下文和对应的目标token
    print(context, "----->", desired)

[290] -----> 4920
[290, 4920] -----> 2241
[290, 4920, 2241] -----> 287
[290, 4920, 2241, 287] -----> 257


In [33]:
# 再次遍历，这次把token解码为可读文本
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]

    # 打印解码后的上下文文本和目标词
    print(tokenizer.decode(context), "----->", tokenizer.decode([desired]))

 and ----->  established
 and established ----->  himself
 and established himself ----->  in
 and established himself in ----->  a


In [34]:
# 导入PyTorch深度学习框架
import torch

# 打印当前PyTorch的版本号，用于确认安装是否成功
print("PyTorch version:", torch.__version__)

PyTorch version: 2.3.1


**我们采用滑动窗口的方式，每次将窗口向右移动 1 个位置：**

![](images/2.13.png)

**创建数据集和数据加载器，从输入文本数据集中提取文本块。**

In [35]:
# 从PyTorch的工具包中导入Dataset和DataLoader，用于构建自定义数据集和数据加载器
from torch.utils.data import Dataset, DataLoader

# 定义一个继承自PyTorch Dataset的自定义数据集类，用于处理GPT模型的训练数据
class GPTDatasetV1(Dataset):
    def __init__(self, txt, tokenizer, max_length, stride):
        # 初始化存储输入和目标token序列的列表
        self.input_ids = []
        self.target_ids = []

        # 将整个输入文本编码为模型可处理的token ID序列
        # allowed_special参数允许包含特殊的结束符"<|endoftext|>"
        token_ids = tokenizer.encode(txt, allowed_special={"<|endoftext|>"})

        # 断言检查：编码后的token数量必须至少比max_length多1
        # 否则无法生成有效的输入-目标对
        assert len(token_ids) > max_length, "编码后的输入数量必须至少等于max_length+1"

        # 使用滑动窗口技术，将长文本切分为多个重叠的max_length长度的序列
        for i in range(0, len(token_ids) - max_length, stride):
            # 截取从i到i+max_length的token作为输入块
            input_chunk = token_ids[i:i + max_length]
            # 截取从i+1到i+max_length+1的token作为目标块（输入向右偏移1位）
            target_chunk = token_ids[i + 1: i + max_length + 1]

            # 将输入块和目标块转换为PyTorch张量，并添加到对应的列表中
            self.input_ids.append(torch.tensor(input_chunk))
            self.target_ids.append(torch.tensor(target_chunk))

    # 返回数据集中样本的总数
    def __len__(self):
        return len(self.input_ids)

    # 根据索引idx获取对应的输入和目标张量对
    def __getitem__(self, idx):
        return self.input_ids[idx], self.target_ids[idx]

In [36]:
def create_dataloader_v1(txt, batch_size=4, max_length=256,
                         stride=128, shuffle=True, drop_last=True,
                         num_workers=0):
    # 初始化分词器：使用tiktoken加载GPT-2的分词器
    tokenizer = tiktoken.get_encoding("gpt2")

    # 创建数据集：使用我们之前定义的GPTDatasetV1类
    dataset = GPTDatasetV1(txt, tokenizer, max_length, stride)

    # 创建数据加载器：将数据集包装成可迭代的批量加载器
    dataloader = DataLoader(
        dataset,
        batch_size=batch_size,      # 每个批次包含的样本数
        shuffle=shuffle,            # 是否在每个epoch打乱数据顺序
        drop_last=drop_last,        # 是否丢弃最后一个不完整的批次
        num_workers=num_workers     # 用于加载数据的子进程数量
    )

    return dataloader

In [37]:
# 打开文本文件并读取内容
with open("the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()

In [38]:
# 创建数据加载器
# 传入原始文本，并设置批量大小为1，上下文窗口长度为4，步幅为1，不打乱数据顺序
dataloader = create_dataloader_v1(
    raw_text, batch_size=1, max_length=4, stride=1, shuffle=False
)

# 将数据加载器转为迭代器，以便逐个获取批次
data_iter = iter(dataloader)
# 获取第一个批次
first_batch = next(data_iter)
# 打印第一个批次的输入和目标张量
print(first_batch)

[tensor([[  40,  367, 2885, 1464]]), tensor([[ 367, 2885, 1464, 1807]])]


In [39]:
# 获取第二个批次
second_batch = next(data_iter)
# 打印第二个批次的输入和目标张量
print(second_batch)

[tensor([[ 367, 2885, 1464, 1807]]), tensor([[2885, 1464, 1807, 3619]])]


In [40]:
# 创建数据加载器
# 设置批量大小为8，上下文窗口长度为4，步幅为4（与窗口长度相同，无重叠），不打乱数据顺序
dataloader = create_dataloader_v1(raw_text, batch_size=8, max_length=4, stride=4, shuffle=False)

# 将数据加载器转为迭代器
data_iter = iter(dataloader)
# 获取第一个批次的输入和目标张量
inputs, targets = next(data_iter)

# 打印输入张量
print("Inputs:\n", inputs)
# 打印目标张量
print("\nTargets:\n", targets)

Inputs:
 tensor([[   40,   367,  2885,  1464],
        [ 1807,  3619,   402,   271],
        [10899,  2138,   257,  7026],
        [15632,   438,  2016,   257],
        [  922,  5891,  1576,   438],
        [  568,   340,   373,   645],
        [ 1049,  5975,   284,   502],
        [  284,  3285,   326,    11]])

Targets:
 tensor([[  367,  2885,  1464,  1807],
        [ 3619,   402,   271, 10899],
        [ 2138,   257,  7026, 15632],
        [  438,  2016,   257,   922],
        [ 5891,  1576,   438,   568],
        [  340,   373,   645,  1049],
        [ 5975,   284,   502,   284],
        [ 3285,   326,    11,   287]])


# 总结

**滑动窗口采样技术**，用于为大语言模型构建 “预测下一个词” 的训练数据，目标序列是输入序列右移 1 个 token 的结果。

关键参数 ** 步幅（stride）** 决定样本重叠程度：步幅等于窗口大小时样本无重叠，可降低过拟合；步幅小于窗口大小时样本重叠，适用于小数据集扩充样本量。

实现了基于 PyTorch 的`GPTDatasetV1`自定义数据集和`create_dataloader_v1`数据加载器，完成文本编码、窗口切分与批量加载。

# 2.7 创建词嵌入

![](images/2.15.png)

- 假设我们有四个输入示例，经过分词后得到的输入 ID 是 2、3、5 和 1：
- 为了简化说明，我们假设词汇表很小，只有 6 个词，并且希望创建维度为 3 的嵌入向量：
- 这会生成一个 6×3 的权重矩阵：

In [41]:
import torch

# 1. 定义输入的token ID（分词后的结果）
# 输入序列包含四个token，对应的ID分别是2、3、5、1
input_ids = torch.tensor([2, 3, 5, 1])

# 2. 定义嵌入层的参数
vocab_size = 6    # 词汇表大小，总共有6个不同的词
output_dim = 3    # 每个词的嵌入向量维度是3

# 3. 设置随机种子，保证每次运行结果一致
torch.manual_seed(123)

# 4. 创建嵌入层
# Embedding层会自动初始化一个 vocab_size × output_dim 的权重矩阵
embedding_layer = torch.nn.Embedding(vocab_size, output_dim)

# 5. 打印嵌入层的权重矩阵
# 这个矩阵里的每一行对应一个词的嵌入向量，会在训练中不断更新
print(embedding_layer.weight)

Parameter containing:
tensor([[ 0.3374, -0.1778, -0.1690],
        [ 0.9178,  1.5810,  1.3010],
        [ 1.2753, -0.2010, -0.1606],
        [-0.4015,  0.9666, -1.1481],
        [-1.1589,  0.3255, -0.6315],
        [-2.8400, -0.7849, -1.4096]], requires_grad=True)


- 嵌入层本质上就是一个**查表（look-up）操作**

In [42]:
# 1. 对单个token ID（3）进行嵌入
# 输入是一个包含ID=3的张量，输出就是权重矩阵中第3行对应的3维向量
print(embedding_layer(torch.tensor([3])))

tensor([[-0.4015,  0.9666, -1.1481]], grad_fn=<EmbeddingBackward0>)


In [43]:
# 2. 对整个输入序列的token IDs进行嵌入
# 输入是[2, 3, 5, 1]，输出是每个ID对应的嵌入向量组成的矩阵
print(embedding_layer(input_ids))

tensor([[ 1.2753, -0.2010, -0.1606],
        [-0.4015,  0.9666, -1.1481],
        [-2.8400, -0.7849, -1.4096],
        [ 0.9178,  1.5810,  1.3010]], grad_fn=<EmbeddingBackward0>)


**嵌入层本质上就是一个查表操作**：

输入文本 `fox jumps over dog` 先被分词并映射为 token ID `[2, 3, 5, 1]`。

嵌入层的权重矩阵是一个 6×3 的表，每个 token ID 对应矩阵里的一行。

当我们需要嵌入时，只需根据 ID 去矩阵中 “查” 对应的行，就能得到该 token 的连续向量表示。

## 嵌入层（Embedding Layer） vs 线性层（Linear Layer）

|     特性     |                  嵌入层（Embedding Layer）                   |                    线性层（Linear Layer）                    |
| :----------: | :----------------------------------------------------------: | :----------------------------------------------------------: |
|   **目标**   | 把离散的类别 / 索引（如词的 ID）映射到低维连续向量空间，捕捉语义关联 | 对连续向量做线性变换（`y = xW + b`），用于特征提取或维度转换 |
| **输入类型** |       离散的整数索引（如 `input_ids = [2, 3, 5, 1]`）        |         连续的浮点向量（如 `x = [0.1, 0.5, -0.3]`）          |
| **典型场景** |             NLP 中的词嵌入、推荐系统中的物品嵌入             | 分类 / 回归任务的输出层、Transformer 中的前馈网络、特征维度变换 |

- **嵌入层**
    - 本质是一个**查找表（lookup table）**，权重矩阵的形状是 `[vocab_size, output_dim]`。
    - 输入一个索引 `i`，就直接取出矩阵的第 `i` 行作为输出，不涉及任何计算。
    - 例如：输入 `input_ids = [2, 3]`，就会取出权重矩阵的第 2 行和第 3 行作为嵌入向量。
- **线性层**
    - 本质是一个**矩阵乘法 + 偏置**运算，权重矩阵的形状是 `[input_dim, output_dim]`。
    - 输入一个向量 `x`，会计算 `y = x @ weight + bias`，是对输入的全局变换。
    - 例如：输入 `x = [1.0, 2.0, 3.0]`，会与权重矩阵做乘法得到输出向量。

# 总结

词元嵌入的作用是将**离散的 token ID**（如分词后的`[2,3,5,1]`）映射为**低维连续向量**，用于大语言模型的输入表示。

嵌入层的本质是**查表操作**：权重矩阵形状为`[vocab_size, output_dim]`，输入 token ID 对应矩阵的行索引，直接取出该行作为嵌入向量。

输入必须是**整数类型的张量**（token ID 序列），输出为对应维度的连续向量矩阵。

嵌入层权重默认可训练（`requires_grad=True`），支持反向传播更新。

# 2.8 编码词的位置

嵌入层会将词 ID 转换为完全相同的向量表示，无论该词在输入序列中的位置如何。

在示例句子 fox jumps over fox 中：

两个 fox 对应的 ID 都是 2

经过嵌入层后，它们得到的向量完全相同（都是 [1.2753, -0.2010, -0.1606]）

这说明普通的嵌入层不包含位置信息，模型无法区分同一个词在句子中的不同位置

位置嵌入会与词嵌入向量组合，形成大语言模型的输入嵌入。

Token embeddings（词嵌入）：只包含词本身的语义信息，同一个词的向量完全相同。

Positional embeddings（位置嵌入）：只包含词在序列中的位置信息，比如第 1 个位置、第 2 个位置的向量各不相同。

Input embeddings（输入嵌入）：将词嵌入和位置嵌入相加，得到的向量同时包含了语义信息和位置信息。

第 1 个 fox 的输入嵌入 = 词嵌入[fox] + 位置嵌入[第1位]

第 4 个 fox 的输入嵌入 = 词嵌入[fox] + 位置嵌入[第4位]

这样模型就能区分这两个 fox 在句子中的不同位置了。

In [44]:
import torch
import torch.nn as nn

# 1. 定义超参数
vocab_size = 10000  # 词汇表大小
embed_dim = 512     # 嵌入向量的维度
max_seq_len = 512   # 最大序列长度

In [45]:
# 2. 定义嵌入层
token_embedding = nn.Embedding(vocab_size, embed_dim)  # 词嵌入层
pos_embedding = nn.Embedding(max_seq_len, embed_dim)   # 位置嵌入层

In [46]:
# 3. 构造输入
# 假设输入序列是 "fox jumps over fox"，对应的ID为 [2, 3, 5, 2]
input_ids = torch.tensor([2, 3, 5, 2])  # 词ID
seq_len = input_ids.size(0)             # 序列长度（这里是4）
positions = torch.arange(seq_len)       # 位置索引 [0, 1, 2, 3]

In [47]:
# 4. 计算嵌入
token_embeds = token_embedding(input_ids)  # 词嵌入，shape: [4, 512]
pos_embeds = pos_embedding(positions)      # 位置嵌入，shape: [4, 512]

In [48]:
# 5. 组合得到输入嵌入
input_embeds = token_embeds + pos_embeds   # 相加，shape: [4, 512]

In [49]:
# 6. 验证效果
print("第一个fox的词嵌入：", token_embeds[0][:5])  # 取前5个维度查看
print("第四个fox的词嵌入：", token_embeds[3][:5])
print("第一个fox的位置嵌入：", pos_embeds[0][:5])
print("第四个fox的位置嵌入：", pos_embeds[3][:5])
print("第一个fox的输入嵌入：", input_embeds[0][:5])
print("第四个fox的输入嵌入：", input_embeds[3][:5])

第一个fox的词嵌入： tensor([ 1.6853, -0.3072, -0.0887,  0.1228,  0.8227], grad_fn=<SliceBackward0>)
第四个fox的词嵌入： tensor([ 1.6853, -0.3072, -0.0887,  0.1228,  0.8227], grad_fn=<SliceBackward0>)
第一个fox的位置嵌入： tensor([-0.1528, -0.7229,  0.4540,  0.1622, -0.1771], grad_fn=<SliceBackward0>)
第四个fox的位置嵌入： tensor([-0.7160,  1.4558, -0.5102,  1.7818,  1.7861], grad_fn=<SliceBackward0>)
第一个fox的输入嵌入： tensor([ 1.5325, -1.0302,  0.3652,  0.2851,  0.6455], grad_fn=<SliceBackward0>)
第四个fox的输入嵌入： tensor([ 0.9693,  1.1486, -0.5990,  1.9046,  2.6088], grad_fn=<SliceBackward0>)


In [60]:
vocab_size = 50257 # GPT 等模型常用的词汇表大小
output_dim = 256 # 嵌入向量的维度为 256

token_embedding_layer = torch.nn.Embedding(vocab_size, output_dim)

当输入 token ID 时，会从这个矩阵中取出对应的行，作为该 token 的向量表示。

如果我们从数据加载器中采样数据，会将每个批次中的 token 嵌入为 256 维向量

如果批次大小为 8，每个样本包含 4 个 token，那么嵌入后的张量形状就是 8 x 4 x 256

In [ ]:
max_length = 4 # 每个样本的最大 token 长度为 4
dataloader = create_dataloader_v1(
    raw_text, batch_size=8, max_length=max_length,
    stride=max_length, shuffle=False
)

In [52]:
data_iter = iter(dataloader) # 将数据加载器转为迭代器
inputs, targets = next(data_iter) # 获取迭代器的下一个元素，得到输入和目标

In [53]:
print("Token IDs:\n", inputs)
print("\nInputs shape:\n", inputs.shape)

Token IDs:
 tensor([[   40,   367,  2885,  1464],
        [ 1807,  3619,   402,   271],
        [10899,  2138,   257,  7026],
        [15632,   438,  2016,   257],
        [  922,  5891,  1576,   438],
        [  568,   340,   373,   645],
        [ 1049,  5975,   284,   502],
        [  284,  3285,   326,    11]])

Inputs shape:
 torch.Size([8, 4])


In [ ]:
token_embeddings = token_embedding_layer(inputs)
print(token_embeddings.shape) # 256 代表每个 token 的嵌入维度；这表示每个批次里的每个 token 都被转换成了一个 256 维的向量。

torch.Size([8, 4, 256])


GPT-2 使用绝对位置嵌入，因此我们只需要再创建一个嵌入层即可。

绝对位置嵌入是给每个位置分配一个固定的向量，模型通过这个向量来区分不同位置的 token。

In [55]:
context_length = max_length # 上下文长度等于最大序列长度（这里是 4）
pos_embedding_layer = torch.nn.Embedding(context_length, output_dim)
# 创建位置嵌入层，输入是位置索引（0 到 3），输出是 256 维向量

In [56]:
pos_embeddings = pos_embedding_layer(torch.arange(max_length)) # 生成位置嵌入，输入是位置索引 [0, 1, 2, 3]
print(pos_embeddings.shape)

torch.Size([4, 256])


要构建大语言模型（LLM）使用的输入嵌入，我们只需要把词嵌入和位置嵌入相加即可。

In [57]:
input_embeddings = token_embeddings + pos_embeddings
print(input_embeddings.shape)

torch.Size([8, 4, 256])


相加后，每个 token 的向量同时包含了语义信息（来自词嵌入）和位置信息（来自位置嵌入），这正是 Transformer 模型需要的输入格式。

# 总结

1. 嵌入层基础：词嵌入层将 token ID 转为固定维度向量（如 256 维），仅包含语义信息，同 token 在不同位置的向量完全相同；GPT-2 采用绝对位置嵌入，通过独立的位置嵌入层为每个位置分配固定向量。
2. 输入嵌入生成：将词嵌入与位置嵌入相加，得到同时包含语义和位置信息的输入嵌入，最终形状为[批次大小, 序列长度, 嵌入维度]（如[8, 4, 256]）。
3. GPT 输入流程：原始文本→切分 token→转 token ID→生成词 / 位置嵌入→相加得到输入嵌入→送入仅解码器的 Transformer 模型→输出转换为可读文本。